### PR TITLE
Remove arc and refcell

### PR DESCRIPTION
**Pull request by RoDmitry**
#### Description:
I've looked at `actix-identity` [middleware](https://github.com/actix/actix-extras/blob/8e35b652d426c0bcb3de44dcf4278320b04761ae/actix-identity/src/middleware.rs) and I think that `arc` and `refcell` is not needed. I might be wrong, I'm still learning, but all tests have passed. And what the `attache` was for? Do you use it elsewhere?

#### Checklist:

- [ ] Tests for the changes have been added (_for bug fixes / features_);
- [ ] Docs have been added / updated (_for bug fixes / features_).
- [ ] This PR has been added to [CHANGELOG.md](https://github.com/DDtKey/actix-web-grants/blob/main/CHANGELOG.md) (to [Unreleased] section);

### CODE REVIEW

File: src/middleware.rs [hunk 1]
```diff
@@ -121,14 +120,13 @@ where
...
-        let extractor = Arc::clone(&self.extractor);
+        let extractor = Rc::clone(&self.extractor);
 
         Box::pin(async move {
             let permissions: Vec<String> = extractor.extract(&req).await?;
-            req.attach(permissions);
 
-            let fut = service.borrow_mut().call(req);
-            fut.await
+            req.extensions_mut().insert(AuthDetails::new(permissions));
```
Comment (DDtKey) [inline src/middleware.rs:128]: `attache` replaces this, allowing users to customize it directly to the request without going into the details of the data structure (`AuthDetails`).
For example when using another middleware, see: https://github.com/DDtKey/actix-web-grants/blob/bd934757fd6d75a8e448bc6733af418d1d70617b/examples/jwt-httpauth/src/main.rs#L43

Comment (RoDmitry) [reply src/middleware.rs]: I reverted it back

**Comment by DDtKey** [comment]
I'm afraid we can't rely on `Rc` since the `extractor` can be used across multiple threads with parallel requests(and there may be a custom implementation of users). 
We clone it with every transformation, see here: https://github.com/DDtKey/actix-web-grants/blob/bd934757fd6d75a8e448bc6733af418d1d70617b/src/middleware.rs#L96

Rather, we should think about tests that will cover this 🤔

**Comment by DDtKey** [comment]
I'm afraid we can't rely on `Rc` since the `extractor` can be used across multiple threads with parallel requests(and there may be a custom implementation of users, not only functions auto-implemented trait `PermissionsExtractor`). 

We clone it with every transformation and request.
It may be unsafe if the counter goes down to zero non-atomic (and resources will be freed).

Rather, we should think about tests that will cover this 🤔

**Comment by RoDmitry** [comment]
I've checked all of the `actix-web/src/middleware/`, and all of them are `Rc`, that's strange 🤔

**Comment by DDtKey** [comment]
> I've checked all of the `actix-web/src/middleware/`, and all of them are `Rc`, that's strange 🤔

It's good point, I want to make sure as it really confuses me

**Comment by RoDmitry** [comment]
_resources will be freed_
I think these are `'static`, and can't be freed

**Comment by DDtKey** [comment]
> _resources will be freed_
> I think these are `'static`, and can't be freed
> It's even stated here:
> 
> https://github.com/DDtKey/actix-web-grants/blob/bd934757fd6d75a8e448bc6733af418d1d70617b/src/middleware.rs#L86

yeah, but `'static` it is not necessarily a static lifetime, it's can be owned value also. And it can be dynamically freed at runtime 🤔 
You can see it in [this one article](https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md#2-if-t-static-then-t-must-be-valid-for-the-entire-program)

**Comment by DDtKey** [comment]
The reason is simpler, single-threaded workers are used in `actix-web`.
That is, the state is not shared between threads, but a new one is created for each one.

So, your suggestion will work well! 👏 
But what do you think about make this for `3+` version ? We will merge it to `v4` anyway

**Comment by RoDmitry** [comment]
Cool! So I just make same on `main`, instead of v4, right?

**Comment by DDtKey** [comment]
> Cool! So I just make same on `main`, instead of v4, right?

Yep! 🙂

**Comment by RoDmitry** [comment]
Implementation for v4 is a bit different

**Comment by RoDmitry** [comment]
Should I change the cargo version to "2.0.2"?

**Comment by DDtKey** [comment]
Thank you for your contribution! 🚀

**Pull request merged**